File: 240704/4가지 연산을 이용하여 1 만들기/make-one-using-four-operations.cpp
#include <iostream>
#include <queue>
#include <algorithm>
using namespace std;

#define MAX_N 1000000
#define OPERATOR_NUM 4

enum OPERATOR {
    SUBTRACT,
    ADD,
    DIV2,
    DIV3
};

int n;
int step[MAX_N*2];
bool visited[MAX_N*2];
queue<int> q;

int ans;

void Push(int num, int new_step) {
    q.push(num);
    visited[num] = 1;
    step[num] = new_step;
}

bool Possible(int num, int op) {
    if(op == SUBTRACT || op == ADD)
        return true;
    else if(op == DIV2)
        return num % 2 == 0;
    else
        return num % 3 == 0;
}

int Calculate(int num, int op) {
    if(op == SUBTRACT)
        return num-1;
    else if (op == ADD)
        return num+1;
    else if (op == DIV2)
        return num/2;
    else
        return num/3;
}

bool InRange(int num) {
    return 1<= num && num <= 2*n-1;
}

bool CanGo(int num) {
    return InRange(num) && !visited[num];
}

void BFS() {
    while(!q.empty()) {
        int curr_num = q.front();
        q.pop();

        for(int i=0; i<OPERATOR_NUM; i++) {
            if(!Possible(curr_num, i)) continue;

            int new_num = Calculate(curr_num, i);

            if(CanGo(new_num))
                Push(new_num, step[curr_num]+1);
        }

    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n;

    Push(n, 0);
    BFS();

    cout << step[1];
    return 0;
}

File: 250206/4가지 연산을 이용하여 1 만들기/make-one-using-four-operations.py
from collections import deque
import sys

OPE_NUM = 4
INT_MAX = sys.maxsize

n = int(input())
ans = INT_MAX

q = deque()
visited = [False for _ in range(2*n)]

step = [0 for _ in range(2*n)]

def possible(num, op):
    if op == 0 or op == 1:
        return True
    elif op == 2:
        return num%2 == 0
    else:
        return num%3 == 0

def calculate(num, op):
    if op == 0:
        return num -1
    elif op == 1:
        return num + 1
    elif op == 2:
        return num//2
    else:
        return num//3

def can_go(num):
    return 1<=num<=2*n-1 and not visited[num]

def find_min():
    global ans
    
    while q:
        cur_num = q.popleft()

        for i in range(OPE_NUM):
            if not possible(cur_num, i):
                continue
            
            new_num = calculate(cur_num, i)

            if can_go(new_num):
                q.append(new_num)
                visited[new_num] = True
                step[new_num] = step[cur_num]+1
        
        ans = step[1]

q.append(n)
visited[n] = True
step[n] = 0

find_min()
print(ans)